// react-colors/lib/components/ColorBlock/ColorBlock.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _getTransparentBackground = require('../../helpers/getTransparentBackground');

var _getTransparentBackground2 = _interopRequireDefault(_getTransparentBackground);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/*function getHue (hsv, num) {
	var i = num, colors = []
	num = num || 12
	while(i > 0){
		colors.push(
			chroma([(hsv[0] + 360 - i*360/num)%360, hsv[1], hsv[2]], 'hsv').hex()
		)
		i--
	}
	return colors
}*/

function ColorBlock(_ref) {
	var style = _ref.style,
	    color = _ref.color,
	    _onClick = _ref.onClick,
	    props = _objectWithoutProperties(_ref, ['style', 'color', 'onClick']);

	var styles = {
		root: Object.assign({}, style, _extends({}, (0, _getTransparentBackground2.default)(color)))
	};
	return _react2.default.createElement('div', _extends({
		style: styles.root,
		onClick: function onClick(e) {
			return _onClick(color, e);
		}
	}, props));
}

ColorBlock.displayName = 'ColorBlock';

ColorBlock.defaultProps = {
	color: _react.PropTypes.string.isRequired
};

exports.default = ColorBlock;
