// react-colors/lib/pickers/Chrome/ChromeDiscPointer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = ChromeDiscPointer;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function ChromeDiscPointer(props) {
  var style = props.style,
      p = _objectWithoutProperties(props, ['style']);

  var root = Object.assign({
    width: '14px',
    height: '14px',
    boxShadow: '0 1px 2px rgba(0,0,0,.32)',
    background: '#f8f8f8',
    borderRadius: '50%',
    transform: 'translate(-5px, -1px)',
    cursor: 'pointer'
  }, style);

  return _react2.default.createElement('div', _extends({ style: root }, p));
}

ChromeDiscPointer.propTypes = {
  style: _react.PropTypes.object
};
