// react-colors/lib/pickers/Chrome/ChromeCirclePointer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = ChromeCirclePointer;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function ChromeCirclePointer(props) {
	var style = props.style,
	    p = _objectWithoutProperties(props, ['style']);

	var root = Object.assign({
		width: '10px',
		height: '10px',
		boxShadow: '0 0 0 1px #fff,0 0 1px 1px rgba(0,0,0,.4)',
		borderRadius: '50%',
		transform: 'translate(-5px, -5px)'
	}, style);

	return _react2.default.createElement('div', _extends({ style: root }, p));
}

ChromeCirclePointer.propTypes = {
	style: _react.PropTypes.object
};
