// react-colors/lib/pickers/Photoshop/PsPointerDoubleTriangle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PsSVGPointer(props) {
	return _react2.default.createElement(
		"svg",
		_extends({ width: "10px", height: "10px", version: "1.1", xmlns: "http://www.w3.org/2000/svg" }, props),
		_react2.default.createElement("path", { d: "M0 5 L6 1 L8 1 A2 2 0 0 1 9 2 L9 8 A2 2 0 0 1 8 9 L6 9 L0 5" })
	);
}

PsSVGPointer.defaultProps = {
	fill: "#fff",
	stroke: "#666"
};

function PsPointerDoubleTriangle() {
	var style = {
		root: {
			position: 'absolute',
			width: 40,
			transform: 'translate(-10px, -5px)'
		},
		left: {
			float: 'left',
			transform: 'rotate(180deg)'
		},
		right: {
			float: 'right'
		}
	};

	return _react2.default.createElement(
		"div",
		{ style: style.root },
		_react2.default.createElement(PsSVGPointer, { style: style.left }),
		_react2.default.createElement(PsSVGPointer, { style: style.right })
	);
}

exports.default = PsPointerDoubleTriangle;
