// react-colors/lib/components/ColorInput/HexInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HexInput = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp, _initialiseProps; /* 颜色属性输入 */

// todo: 添加propTypes和defaultProps
// todo: debug拖动改数功能
// todo: 客制化标签
// todo: 配置小数点位数
// todo: Lab值变化bug
// todo: fix cmyk.k
// todo: 把number input抽象出来，number input组件接口为通用接口，可以使用任意其它组件替换，包裹到colorInput中都可以使用
// todo: this === active 时，不merge值

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _padStart = require('../../helpers/padStart');

var _padStart2 = _interopRequireDefault(_padStart);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var isStrict16 = function isStrict16(v) {
  var hexRgx = /^[0-9a-fA-F]{0,9}$/;
  v = v.replace('#', '');
  return hexRgx.test(v);
};
var formatHex = function formatHex(hexStr) {
  var max = 16777215;
  var min = 0;
  hexStr = hexStr.replace('#', '');
  var num = parseInt(hexStr, 16);
  var newValue = num > max ? 'ffffff' : hexStr;
  newValue = num < min ? '000000' : newValue;
  return '#' + (0, _padStart2.default)(newValue, 6, '0');
};
var _16To10 = function _16To10(v) {
  return parseInt(v, 16);
};
var _10To16 = function _10To16(v) {
  return Number(v).toString(16);
};

var HexInput = exports.HexInput = (_temp = _class = function (_ref) {
  _inherits(HexInput, _ref);

  function HexInput(props) {
    _classCallCheck(this, HexInput);

    var _this = _possibleConstructorReturn(this, (HexInput.__proto__ || Object.getPrototypeOf(HexInput)).call(this));

    _initialiseProps.call(_this);

    var newValue = props.value.replace('#', '');
    if (props.sharp) newValue = '#' + newValue;
    _this.state = {
      value: newValue,
      blurValue: null
    };
    return _this;
  }

  _createClass(HexInput, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var input = this.input;
      var newValue = nextProps.value.replace('#', '');
      if (nextProps.sharp) newValue = '#' + newValue;

      if (newValue !== this.state.value) {
        if (input === document.activeElement) {
          this.setState({ blurValue: newValue });
        } else {
          this.setState({ value: newValue });
        }
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this,
          _React$createElement;

      // todo 后部标签
      var _props = this.props,
          style = _props.style,
          model = _props.model;


      return _react2.default.createElement('input', (_React$createElement = {
        style: style,
        ref: function ref(node) {
          return _this2.input = node;
        },
        value: this.state.value
      }, _defineProperty(_React$createElement, 'style', style), _defineProperty(_React$createElement, 'onKeyDown', this.handleKeyDown), _defineProperty(_React$createElement, 'onChange', function onChange(e) {
        return _this2.handleChange(e.target.value, e);
      }), _defineProperty(_React$createElement, 'onBlur', this.handleBlur), _defineProperty(_React$createElement, 'placeholder', this.props.placeholder), _React$createElement));
    }
  }]);

  return HexInput;
}(_react.PureComponent || _react.Component), _class.propTypes = {
  style: _react.PropTypes.object,
  value: _react.PropTypes.string.isRequired,
  step: _react.PropTypes.number,
  sharp: _react.PropTypes.bool,
  onChange: _react.PropTypes.func
}, _class.defaultProps = {
  label: _react2.default.createElement(
    'span',
    null,
    '#'
  ),
  step: 1,
  sharp: false,
  onChange: function onChange() {}
}, _initialiseProps = function _initialiseProps() {
  var _this3 = this;

  this.handleBlur = function () {
    if (_this3.state.blurValue) {
      _this3.setState({ value: _this3.state.blurValue, blurValue: null });
    }
  };

  this.handleChange = function (newValue, e, call) {

    if (isStrict16(newValue) || !newValue || newValue == '#') {
      _this3.props.onChange(formatHex(newValue), e);
      _this3.setState({ value: newValue }, call);
    }
  };

  this.handleKeyDown = function (e) {
    var step = _this3.props.step;

    var newValue = e.target.value.replace('#', '');
    var num = parseInt(newValue, 16);
    // Up
    if (e.keyCode === 38) {
      e.preventDefault();
      _this3.handleChange(formatHex(_10To16(num + step)), e, function () {
        return _this3.input.select();
      });
    }

    // Down
    if (e.keyCode === 40) {
      e.preventDefault();
      _this3.handleChange(formatHex(_10To16(num - step)), e, function () {
        return _this3.input.select();
      });
    }
  };
}, _temp);
exports.default = HexInput;
